File: Play/kafka/go-examples/producer.go
package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strconv"
	"time"

	"github.com/IBM/sarama"
)

// Message represents the data structure for Kafka messages
type Message struct {
	ID        int    `json:"id"`
	Message   string `json:"message"`
	Timestamp int64  `json:"timestamp"`
}

func main() {
	// Get parameters
	topic := "test-topic"
	numMessages := 10

	if len(os.Args) > 1 {
		topic = os.Args[1]
	}
	if len(os.Args) > 2 {
		if n, err := strconv.Atoi(os.Args[2]); err == nil {
			numMessages = n
		}
	}

	// Kafka configuration
	config := sarama.NewConfig()
	config.Producer.Return.Successes = true
	config.Producer.RequiredAcks = sarama.WaitForAll
	config.Producer.Retry.Max = 5

	// Create producer
	producer, err := sarama.NewSyncProducer([]string{"localhost:9092"}, config)
	if err != nil {
		log.Fatalf("❌ Failed to create producer: %v", err)
	}
	defer producer.Close()

	fmt.Printf("📤 Starting to produce %d messages to topic '%s'...\n\n", numMessages, topic)

	// Produce messages
	for i := 0; i < numMessages; i++ {
		msg := Message{
			ID:        i,
			Message:   fmt.Sprintf("Hello Kafka message #%d", i),
			Timestamp: time.Now().Unix(),
		}

		// Serialize to JSON
		msgBytes, err := json.Marshal(msg)
		if err != nil {
			log.Printf("❌ Failed to marshal message: %v", err)
			continue
		}

		// Create Kafka message
		kafkaMsg := &sarama.ProducerMessage{
			Topic: topic,
			Key:   sarama.StringEncoder(strconv.Itoa(i)),
			Value: sarama.ByteEncoder(msgBytes),
		}

		// Send message
		partition, offset, err := producer.SendMessage(kafkaMsg)
		if err != nil {
			log.Printf("❌ Failed to send message: %v", err)
		} else {
			fmt.Printf("✅ Message delivered to %s [%d] at offset %d\n", topic, partition, offset)
		}

		time.Sleep(500 * time.Millisecond)
	}

	fmt.Println("\n✨ All messages sent successfully!")
}

File: Play/kafka/python-examples/producer.py
#!/usr/bin/env python3
"""
Simple Kafka Producer Example
Sends messages to a Kafka topic
"""

from confluent_kafka import Producer
import json
import time
import sys

# Kafka configuration
config = {
    'bootstrap.servers': 'localhost:9092',
    'client.id': 'python-producer'
}

# Create Producer instance
producer = Producer(config)


def delivery_callback(err, msg):
    """Callback called once message is delivered or failed"""
    if err:
        print(f'❌ Message delivery failed: {err}')
    else:
        print(f'✅ Message delivered to {msg.topic()} [{msg.partition()}] at offset {msg.offset()}')


def produce_messages(topic='test-topic', num_messages=10):
    """Produce messages to Kafka topic"""
    print(f"📤 Starting to produce {num_messages} messages to topic '{topic}'...\n")

    try:
        for i in range(num_messages):
            # Create message
            message = {
                'id': i,
                'message': f'Hello Kafka message #{i}',
                'timestamp': time.time()
            }

            # Produce message
            producer.produce(
                topic=topic,
                key=str(i),
                value=json.dumps(message),
                callback=delivery_callback
            )

            # Trigger delivery reports
            producer.poll(0)

            time.sleep(0.5)

        # Wait for all messages to be delivered
        print("\n⏳ Flushing remaining messages...")
        producer.flush()
        print("✨ All messages sent successfully!")

    except KeyboardInterrupt:
        print("\n⚠️  Interrupted by user")
    except Exception as e:
        print(f"❌ Error: {e}")
    finally:
        producer.flush()


if __name__ == '__main__':
    topic = sys.argv[1] if len(sys.argv) > 1 else 'test-topic'
    num_messages = int(sys.argv[2]) if len(sys.argv) > 2 else 10

    produce_messages(topic, num_messages)
